certora/specs/erc4626.spec

```diff
@@ -0,0 +1,790 @@
+    uint256 assets2;
+    uint256 assets3;
+    require assets1 < ATokAllowance1;
+    uint256 previewShares1 = previewDeposit(assets1);
+
+    env e1;
+    address receiver1;
+    deposit(e1, assets2, receiver1);
+    
+    uint256 ATokAllowance2 = ATok.allowance(currentContract, user);
+    require assets1 == ATokAllowance2;
+    uint256 previewShares2 = previewDeposit(assets1);
+    
+    env e2;
+    address receiver2;
+    deposit(e2, assets3, receiver2); 
+    
+    uint256 ATokAllowance3 = ATok.allowance(currentContract, user);
+    require assets1 > ATokAllowance3;
+    uint256 previewShares3 = previewDeposit(assets1);
+
+    assert previewShares1 == previewShares2,"previewDeposit should not change regardless of assets > or = allowance";
+    assert previewShares2 == previewShares3,"previewDeposit should not change regardless of assets < or = allowance";
+}
+
+
+// Since maxDeposit is a constant, it cannot have any impact on the previewDeposit value.
+// STATUS: Verified for all f except metaDeposit which has a reachability issue
+// https://vaas-stg.certora.com/output/11775/42e457472f3eeb222744/?anonymousKey=d8e2a760f475fde4ccd07c0bbc4f5703887b9280
+rule maxDepositConstant(method f){
+    address receiver;
+    uint256 maxDep1 = maxDeposit(receiver);
+    calldataarg args;
+    env e;
+    f(e, args);
+    uint256 maxDep2 = maxDeposit(receiver);
+
+    assert maxDep1 == maxDep2,"maxDeposit should not change";
+
+    // address receiver;
+    // uint256 maxDepositAssets1 = maxDeposit(receiver);
+    // uint256 assets;
+    // assert false;
+    // require assets = maxDepositAssets1;
+    // uint256 previewShares1 = previewDeposit(assets);
+    
+    // calldataarg args;
+    // env e1;
+    // f(e1, args);
+    // uint256 maxDepositAssets2 = maxDeposit(receiver);
+    
+    // // require assets <= maxDepositAssets2;
+    // uint256 previewShares2 = previewDeposit(assets);
+    
+    // assert previewShares2 == previewShares1,"previewDeposit should be independent of maxDeposit";
+}
+
+/***
+* rule to check the following for the previewMint function:
+* _1. MUST return as close to and no fewer than the exact amount of assets that would be deposited in a mint call in the same transaction. I.e. mint should return the same or fewer assets as previewMint if called in the same transaction.
+* _2. MUST NOT account for mint limits like those returned from maxMint and should always act as though the mint would be accepted, regardless if the user has enough tokens approved
+*  a. Check against maxDeposit
+*  b. Check against allowance
+* 3. MUST be inclusive of deposit fees. Integrators should be aware of the existence of deposit fees.
+* 4. MUST NOT revert due to vault specific user/global limits
+*/
+// STATUS: Verified, that the amount returned by previewDeposit is exactly equal to that returned by the deposit function.
+//         This is a stronger property than the one required by the EIP.
+// Need to investigate the revert behaviour
```

Comment:
Indeed all revert-related checks are commented out. Let's communicate to the customer that we don't check revert-related properties.